### ISSUE TITLE
ttnn.log accuracy issue

### ISSUE THREAD

**Issue by djordje-tt** [tenstorrent/tt-metal#33775]
### Issue description

While working on adding new feature (log-probs) to TT-Transformers I noticed that after latest rebase I am observing pcc failure on unit test with specific inputs. Since inputs are logits and majority of log-probs values do ttnn.log(exp_sum) where exp_sum is e^(negative numbers mostly) exp_sum becomes close to 1 (~1.002305) when operated with ttnn.log(1.002305) = 0.0010 but with new `ttnn.log` implementation I am getting 0 as final output. Unfortunately numbers are very small and even when converted to `ttnn.float32` and got back from device to host it sees them like zeros and for that matter I needed to use whole test instead of just setting torch inputs, moving to device and running ttnn.log against it for creating repro. 

commit that introduced bug: https://github.com/tenstorrent/tt-metal/commit/de62c46d07e0b3179aaae39848b85eb48a7a62ea

### Setup
machine=`T3K`
command: `pytest tests/ttnn/unit_tests/operations/eltwise/test_sampling.py -k test_log_probs_calculation`
branch_with_failure: `divanovic/add_log_probs_with_penalties_log_bug_repro`
same as prev branch + ttnn.log commit revert: `divanovic/add_log_probs_with_penalties_log_bug_repro_with_revert`

### Repro steps
- upload `qwen_logits.pt` and `qwen_indices.pt` to tt-metal dir. They are compressed and attached here: [inputs.zip](https://github.com/user-attachments/files/23931759/inputs.zip)
```
git checkout divanovic/add_log_probs_with_penalties_log_bug_repro
git submodule update --init --recursive
./create_venv.sh
./build_metal.sh
pytest tests/ttnn/unit_tests/operations/eltwise/test_sampling.py -k test_log_probs_calculation
```

### Observed behaviour
Since ttnn.log produces zeros here pcc=0.0 with all zeros in output tensor

### Expected behaviour pcc=1.0

Let me know if you have any questions!
P0: We are required to deliver log-probs feature to UnsungFields and for that matter we need to fix this asap!

**Comment by nmauriceTT** [comment]
After talking with @djordje-tt (feel free to correct me if I made any mistake) , we made the following observations:

This problems is on T3K, which is Wormhole (I found a bug ttnn.log but it is unrelated, and only on Blackhole for float32).

The input of logarithm is a bfloat16 tensor that is full of 1. (`1.002305` can not be exactly represented with bfloat16 and gets rounded to 1).
Before, ttnn.log(1) = 0.0010
Now, ttnn.log(1) = 0 (correct result)

The expected result (golden output of torch on float32) is a tensor with values ranging from (9e-6 to 1.2e-4). 
In absolute terms, the new output with the new log is closer. 
Previous test likely passed because PCC of constant tensor is set to 1 (expected input 1.0025 rounds to this whwen using bfloat16).

Given these observations, I believe that the new ttnn.log implementation did not cause the intrinsic problem, but instead revealed it (inaccurate of previous ttnn.log was hiding it)

Djorde switched the data type to float32 for all operations in the log probs calculation but it did not improve the result (log was still getting 1 as input and thus giving 0 as output).

From what I can see, the other operations that are involved are:
- ttnn.exp
- ttnn.all_gather
- ttnn.sum
- ttnn.substract

In this specific case, I am suspecting either ttnn.sum (accumulation errors, especially if using bfloat16), or subtract (if done on FPU and truncating inputs to bfloat16).

I can take a look at `ttnn.subtract` to check how accurate it is. 
(EDIT: ttnn.subtract seem accurate within 1 ULP on both bfloat16 and torch32)


Another suggestion would be compute the non-stable 'log-softmax' (`log(X) = exp(x_i) / Sum(exp(x_i)`) (i.e. not subtract maximum). This is less numerically stable but may make increase the orders of magnitudes (?)

**Comment by zzigler-tt** [comment]
@nmauriceTT what's the status of this issue, it is high priority.

**Comment by aczajkowskiTT** [comment]
@ntarafdar Based on discussion about the error it is coming from all_gather as part of the input.  Due to the recent fix in accuracy of log, this problem was hidden before as log was producing erroneous output around the test input.  Now that accuracy is fixed here, the issue has moved upstream.  Since this is going in CCLs, please advise on correct owner.

**Comment by djordje-tt** [comment]
@ekuslysTT I am removing C1 since this is not a blocker and does not affect our release for customer. And this probably needs further investigation which op is causing this for ttnn.float32. For bfloat16 this is expected due to precision's inability to capture very close numbers around 1.
